src/Simulation/Native/src/simulator/wavefunction.hpp

```diff
@@ -90,6 +91,52 @@ class Wavefunction
 
     void flush() const
     {
+        //logic to reorder
+        Fusion fg = fused_.get_fusedgates();
+        auto itemsToFuse = fg.get_items();
+        auto ctrlSet = fg.get_ctrl_set();
+        //getting all qubits to move to lower end of the wfn
+        if (itemsToFuse.size() > 0) {
+            std::vector<unsigned> unionOfAllQubitsInUse;
+            std::unordered_set<unsigned> indicesSet; //set is introduced to guard against duplicate insertion and maintianing original order
+            for (int i = 0; i < itemsToFuse.size(); i++) {
+                auto tempIndices = itemsToFuse[i].get_indices();
+                for (unsigned j = 0; j < tempIndices.size(); j++) {
+                    if (indicesSet.count(tempIndices[j]) == 0) {
+                        unionOfAllQubitsInUse.push_back(tempIndices[j]);
+                        indicesSet.insert(tempIndices[j]);
+                    }
+                }
+            }
+            for (auto it = ctrlSet.begin(); it != ctrlSet.end(); ++it) {
+                if (indicesSet.count(*it) == 0) {
+                    unionOfAllQubitsInUse.push_back(*it);
+                    indicesSet.insert(*it);
+                }
+            }
+            //performing reorder
+            std::vector<qubit_t> indexLocs = qubits(unionOfAllQubitsInUse);
+            for (unsigned i = 0; i < indexLocs.size(); i++)
+            {
+                auto currLoc = indexLocs[i];
+                reorder_wavefunction(currLoc, i);
+                indexLocs = qubits(unionOfAllQubitsInUse);
```

Comment:
is this repeated assignment inside the loop necessary?

Reply:
The reason we cannot do a single update after the loop is because every iteration of the reorder routine changes the mapping, and if this updated mapping is not used for each iteration it results in buggy logic.

Reply:
ah, yes, you are reading from the updated indexLocks inside the loop... 

The code is extremely hard to understand and britle. For example, going back to the orderding questions I had above: the loop progresses through indexLocs array even though its content keeps changing -- what does it mean for the final reordering? what state will we be in for variours possible inputs? which ones of them are a win for perf and which aren't? can we end up reordering the same qubit back and forth?